MainWindow.cpp

```diff
@@ -49,7 +63,88 @@ MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent), ui(new Ui::MainWi
 
 MainWindow::~MainWindow() { delete ui; }
 
-void MainWindow::closeEvent(QCloseEvent *event) { event->accept(); }
+void MainWindow::readSettings() {
+  QSettings settings;
+
+  settings.beginGroup("MainWindow");
+  restoreGeometry(settings.value("geometry").toByteArray());
+  restoreState(settings.value("state").toByteArray());
+  settings.endGroup();
+}
+
+void MainWindow::writeSettings() {
+  QSettings settings;
+
+  settings.beginGroup("MainWindow");
+  settings.setValue("geometry", saveGeometry());
+  settings.setValue("state", saveState());
+  settings.endGroup();
+}
+
+static inline QString recentFilesKey() { return QStringLiteral("recentFileList"); }
+static inline QString fileKey() { return QStringLiteral("file"); }
+
+bool MainWindow::hasRecentFiles() {
+  QSettings settings;
+
+  const int count = settings.beginReadArray(recentFilesKey());
+  settings.endArray();
+  return count > 0;
+}
+
+static QStringList readRecentFiles(QSettings &settings) {
+  QStringList result;
+  const int count = settings.beginReadArray(recentFilesKey());
+  for (int i = 0; i < count; ++i) {
+    settings.setArrayIndex(i);
+    result.append(settings.value(fileKey()).toString());
+  }
+  settings.endArray();
+  return result;
+}
+
+void MainWindow::updateRecentFileActions() {
+  QSettings settings;
+
+  const QStringList recentFiles = readRecentFiles(settings);
+  const int count = qMin(int(MaxRecentFiles), recentFiles.size());
+  int i = 0;
+  for (; i < count; ++i) {
+    const QString fileName = QFileInfo(recentFiles.at(i)).fileName();
+    QString numberString = QString::number(i + 1);
+    recentFileActs[i]->setText(numberString.insert(numberString.length() - 1, '&') + " " + fileName);
+    recentFileActs[i]->setData(recentFiles.at(i));
+    recentFileActs[i]->setVisible(true);
+  }
+  for (; i < MaxRecentFiles; ++i) recentFileActs[i]->setVisible(false);
+}
+
+static void writeRecentFiles(const QStringList &files, QSettings &settings) {
+  const int count = files.size();
+  settings.beginWriteArray(recentFilesKey());
+  for (int i = 0; i < count; ++i) {
+    settings.setArrayIndex(i);
+    settings.setValue(fileKey(), files.at(i));
+  }
+  settings.endArray();
+}
+
+void MainWindow::prependToRecentFiles(const QString &fileName) {
+  QSettings settings;
+
+  const QStringList oldRecentFiles = readRecentFiles(settings);
+  QStringList recentFiles = oldRecentFiles;
+  recentFiles.removeAll(fileName);
+  recentFiles.prepend(fileName);
+  if (oldRecentFiles != recentFiles) writeRecentFiles(recentFiles, settings);
+
+  this->ui->menuRecent->setEnabled(!recentFiles.isEmpty());
+}

```

Comment:
You could move all these "recent files" methods to a `RecentFiles.cpp` and miss nothing for it.

Reply:
Moved with the other logic for recent files to `Components/RecentFiles.cpp`